akka-stream/src/main/scala/akka/stream/stage/Stage.scala

```diff
@@ -28,9 +29,30 @@ import akka.stream.Supervision
 sealed trait Stage[-In, Out]
 
 private[stream] abstract class AbstractStage[-In, Out, PushD <: Directive, PullD <: Directive, Ctx <: Context[Out]] extends Stage[In, Out] {
-  private[stream] var holding = false
-  private[stream] var allowedToPush = false
+  /**
+   * INTERNAL API
+   */
+  private[stream] var holdingUpstream = false
+  /**
+   * INTERNAL API
+   */
+  private[stream] var holdingDownstream = false
+  /**
+   * INTERNAL API
+   */
+  private[stream] var precedingWasPull = false
+  /**
+   * INTERNAL API
+   */
   private[stream] var terminationPending = false
```

Comment:
no, it is unbounded recursion now :) There is a comment above somewhere where we discuss using bitfields, not because of memory but because it allows cheaper tests where we actually use these.